### PR TITLE
Jupyter Caddyfile

### PR DESCRIPTION
**Pull request by jcrubino**
This is a Caddyfile for Jupyter notebooks.

### CODE REVIEW

File: jupyter/Caddyfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+ 
+  proxy / localhost:8888 {
+    header_upstream X-Real-IP  {remote}
+    header_upstream Host {host}
+    websocket
+  }
+
+    rewrite {
+        r "~* /(api/kernels/[^/]+/(channels|iopub|shell|stdin)|terminals/websocket)/?"
```
Comment (mholt) [inline jupyter/Caddyfile]: Have you tested this expression? I'm not familiar with the `~* ` prefix. What does it do?

**Comment by mholt** [comment]
@jcrubino I don't think that syntax works in Go. Try it here: https://regex-golang.appspot.com/assets/html/index.html

Go uses RE2 syntax, described here succinctly: https://github.com/google/re2/wiki/Syntax

So, once this is revised, give it a try, push your change, and I'll take another look! :)

**Comment by jcrubino** [comment]
Using the regex given in the Caddyfile example for this pull request does
not perform a match in the appspot tool you linked to.  Lets call this
regex A.

Using "~*/(api/kernels/[^/]+/(channels|iopub|shell|stdin)|terminals/websocket)/?"
in the appspot tool you linked to performs a match. Let's call this regex B.

Regex A works with the jupyter notebook. UI connects, websocket chatter
connects.

Regex B will allow Jupyter Notebook UI through (mostly), but the websocket
chatter does not connect (processing kernel).

The Tilde operater seems to be a defacto regex operator for column
processing which applies to url/uri processing.
It also appears to be un-documented in the Re2, Golang, and most Posix
Regex References.

We have the choice between using a working bike shed, or stopping the use
of the bike shed for lumber realignment and paint re-evaluation.  : )

**Comment by mholt** [comment]
@jcrubino What is the URL that doesn't match in one but does with the other?

**Comment by jcrubino** [comment]
Updated using Caddyfile posted in caddy issues [here](https://github.com/mholt/caddy/issues/763)

**Comment by mholt** [comment]
It would be good if you could test this first. Let us know if it works!

**Comment by jcrubino** [comment]
Caddyfile is tested and works.
If there is a standardized test suite available, let me know.

**Comment by mholt** [comment]
How sure are you? `proxy_header` hasn't been a valid subdirective since it was replaced almost a year ago by another one. You should be getting an error when you run this Caddyfile.

**Comment by jcrubino** [comment]
You are correct.
My local editor was using an old remote file location, not the current file
location.
The repo update is tested and uses the current Caddyfile in master.

**Pull request merged**